### ISSUE TITLE
Cramtools-3.0.jar error

### ISSUE THREAD

**Issue by jkbonfield** [enasequence/cramtools#20]
I don't seem able to create any CRAM files using the cramtools-3.0.jar.  I checked out the latest "cram3" branch and rebuilt.  I get this error:

```
java -Xmx4000m -jar ~/work/cram/cramtools-3.0.jar cram -R /nfs/srpipe_references/references/Human/1000Genomes_hs37d5/all/fasta/hs37d5.fa --capture-all-tags -Q -n -I _.sam -O _.cram
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at net.sf.cram.CramTools.invoke(CramTools.java:93)
        at net.sf.cram.CramTools.main(CramTools.java:123)
Caused by: java.lang.NullPointerException
        at net.sf.cram.encoding.writer.Writer.write(Writer.java:185)
        at net.sf.cram.build.ContainerFactory.buildSlice(ContainerFactory.java:199)
        at net.sf.cram.build.ContainerFactory.buildContainer(ContainerFactory.java:89)
        at net.sf.cram.build.ContainerFactory.buildContainer(ContainerFactory.java:58)
        at net.sf.cram.Bam2Cram.main(Bam2Cram.java:433)
        ... 6 more
```

Vadim, please email me for an example if you need one,.

**Comment by vadimzalunin** [comment]
what's in the sam file?

**Comment by jkbonfield** [comment]
On Tue, Apr 07, 2015 at 06:13:48AM -0700, Vadim Zalunin wrote:
> what's in the sam file?

Attached, maybe this'll work in github via email.

-- 
James Bonfield (jkb@sanger.ac.uk) | Hora aderat briligi. Nunc et Slythia Tova
                                  | Plurima gyrabant gymbolitare vabo;
  A Staden Package developer:     | Et Borogovorum mimzebant undique formae,
https://sf.net/projects/staden/   | Momiferique omnes exgrabure Rathi. 



-- 
 The Wellcome Trust Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE. 
@HD	VN:1.4	GO:none	SO:coordinate
@RG	ID:H06HD.1	PL:illumina	PU:H06HDADXX130110.1.ATCACGAT	LB:Solexa-135852	SM:NA12878	CN:BI
@RG	ID:H06HD.2	PL:illumina	PU:H06HDADXX130110.2.ATCACGAT	LB:Solexa-135852	SM:NA12878	CN:BI
@RG	ID:H06JU.1	PL:illumina	PU:H06JUADXX130110.1.ATCACGAT	LB:Solexa-135852	SM:NA12878	CN:BI
@SQ	SN:1	LN:249250621	M5:1b22b98cdeb4a9304cb5d48026a85128	UR:/nfs/srpipe_references/references/Human/1000Genomes_hs37d5/all/fasta/hs37d5.fa
H06HDADXX130110:2:2203:6581:76446	83	1	9997	0	190S60M	=	10016	73	CCTGGTGTGCCACGTGGCTCCAGCGCGTACTGTCACGTCAGACTGCGCCTCATTACTACTTCTAGTAGTACACGAAACACGACCCCGTAACCTAACTCTTGTTTTAACTCGGACTCTAACCCTTGCTCTTTCCCAAGCCCTAGCCCGAACACGAGCCCGAGATCTAACTCGAGCTCTGACGCGAGCTCTACCGATAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTA	#####################################################################################################################################################################################################?>???>>????>>>????>>???>>>???>>>???@>>>?>>????@?>>;>=	BD:Z:IIIINNMMMLLKKKKKKJJJJJJJJJJJIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHLKIIKHLKIIKHLKIIKHLKIIKHLKIIKHLKIIKHLLJJLINMLLOLMLII	BI:Z:KKKKIIJLMNNNNNNNMMMMMMMMMMMLLLLLLLLLLLLLLLLLLLLLLKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKJJJKKKKKKKKKKKKKKKKKKKKKKKKKKKKJKKKKKKKKKKKKKKKKKKKKKKKKKKKKJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJMLLKMJMLLKMJMLLKMJMLLKMJMLLKMJMLLKMJMLLKMKNMMMOMNMKK	BQ:Z:@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@>>>>>>>ZYZZZYYZZZZYYYZZZZYYZZZYYYZZZYYYZZZ[YYYZYYZZZZ[ZYYVYX	MQ:i:0	AS:i:60	XS:i:56	RG:Z:H06HD.2
H06JUADXX130110:1:1108:6424:55322	83	1	9998	0	138S107M5S	=	10002	107	TCGTTGCCCTTTACAGATCTCGAACCGTTTCCCTACCTTAATTCCTTGACTTCTTTTTAACCCCATCTCCATCCCTATCTCTTTCCCTGACTCTAACCCTACCCATAGCCCTAACTCTATCCCTATCCCTATCCCTATCGATAACCCTAACCCTAACCCTAACCCTAACCCTATCCCTATCCCTAACCCTAGCCCTATCCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTATCCCT	######################################################################################################################################################################################@?98>?@?A+;?>:;*;8??<=???????@@?????>?@>@>>?@>????@????@@??@:?9*>;>>	BD:Z:IIIINNMMLLLKKKKKKJJJJJJJJJJJJJIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHKHIKHLKJLMHLKFJHHLKHIKHLKHIKHLKHIKHLKHIKHLKHIKHLKHIKHMLIJLJNNINLIMII	BI:Z:KKKKJIKLMNNNNNNNNMMMMMMMMMMMLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKJKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKMMLNJMLMONJMLHMKJMLLKNJMLLKNJMLLKNJMLLKNJMLLKNJMLLKNJMLLLOKNNJONKNKK	BQ:Z:@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>[ZTSYZ[Z\FVZYUVEVSZZWXZZZZZZZ[[ZZZZZYZ[Y[YYZ[YZZZZ[ZZZZ[[ZZ[UZT@@@@@	MQ:i:0	AS:i:87	XS:i:84	RG:Z:H06JU.1

**Comment by jkbonfield** [comment]
It's caused by the ??:i:* tags.  If I remove the 3 integer tags then it can encode OK.

**Comment by vadimzalunin** [comment]
Fixed.
The bug was due to a typo: used external block id instead of tag id when creating codec. Insufficient test coverage... Please give it a go.

**Comment by jkbonfield** [comment]
It fixed the encoder, but it can no longer decode my larger example file:

Failed at record 113. Here is the previously read record: [H06HDADXX130110:2:1110:1498:95067; flags=133; aloffset=0; mateoffset=-1; mappingQuality=0; bases: TGAAAAGGGCAAAACCGGGGGTAATGACATGAGGGTTAAAGTTTGAGTGAGGGTTTTTGCTGCAAGAGGGGATATGCATGCTGCAGGCAGGGGATAGCGGGTCGGGTGACGCCGGGCTGGCCATGGGGTGAAGGCGGGCGCGCCAGAAAGAGGTGCCGAGCAGAAGGTGGGTCGAGCACGAACGGATGTGGTTGCGCGAGGCCGCATCATGCGNGGCGATCCAAGCTCGNNNGGNTANCNNGNGNNTGNN; qscores: ]
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at net.sf.cram.CramTools.invoke(CramTools.java:93)
        at net.sf.cram.CramTools.main(CramTools.java:123)
Caused by: java.lang.RuntimeException: java.nio.BufferUnderflowException
        at net.sf.cram.encoding.reader.CramRecordReader.read(CramRecordReader.java:203)
        at net.sf.cram.build.ContainerParser.getRecords(ContainerParser.java:117)
        at net.sf.cram.build.ContainerParser.getRecords(ContainerParser.java:162)
        at net.sf.cram.build.ContainerParser.getRecords(ContainerParser.java:60)
        at net.sf.cram.Cram2Bam.main(Cram2Bam.java:235)
        ... 6 more
Caused by: java.nio.BufferUnderflowException
        at java.nio.Buffer.nextGetIndex(Buffer.java:492)
        at java.nio.HeapByteBuffer.get(HeapByteBuffer.java:135)
        at net.sf.cram.structure.ReadTag.readSingleValue(ReadTag.java:379)
        at net.sf.cram.structure.ReadTag.restoreValueFromByteArray(ReadTag.java:192)
        at net.sf.cram.structure.ReadTag.<init>(ReadTag.java:57)
        at net.sf.cram.encoding.reader.CramRecordReader.read(CramRecordReader.java:98)
        ... 10 more


Scramble also cannot decode this file.  I'll upload some test data to ngs.sanger.

**Comment by vadimzalunin** [comment]
the fix 494994fc9aa900094024bb5f0c4105658d9e5b82
